### PR TITLE
Implement FailureConverter API

### LINKED ISSUE #1553: FailureConverter API
Add public `FailureConverter` API to allow customizing error <-> failure conversion and `DefaultFailureConverter` implementation that supports encrypting common failure attributes (`message`, `stack_trace`).

This API was added to the other 3 major SDKs already.

### PR DESCRIPTION
**Pull request by mjameswh**
## What was changed

- Exception/Failure conversions now goes through the `DataConverter`;
- The Exception/Failure conversion process can now be customized by passing a custom implementation of the `FailureConverter` interface to DefaultDataConverter;
- The `PayloadDataConverter` now supports encoding of Failure's `message` and `stackTrace` attributes. This is an opt-in feature, that can be enabled by setting `encodeFailureAttributes = true` on constructor.

## Why?

- This API was added to the other 3 major SDKs already.
- Fix #1553

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (mjameswh) [inline temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:133]: @bergundy Can you please confirm that `Failure` details also have to be codec encoded, and that it is _non-optional_ (provided that codecs have indeed been registered)? That means that updated clients that are configured with codecs may suddenly be unable to decode failure details from history generated by previous workers.

Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:133]: > That means that updated clients that are configured with codecs may suddenly be unable to decode failure details from history generated by previous workers.

That's not how Codecs should be implemented though. If implemented correctly. Codecs should check the `encoding` field and don't try to decode that they didn't encode in the first place.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (mjameswh) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:120]: I'm hesitant here on either this method (and the equivalent one below) should be `protected`, so that users may more easily redefine only the "decision chain", while reusing the `encoded_attributes` and stack trace bits below.

On the opposite side, we could make this class `final` and encourage customization by decorator paradigm instead.

I personally have a slight preference for the second approach, for it promote a safer customization mechanism for the vast majority of users (ie. users are more likely to convert their custom exceptions to Temporal exceptions). I however expect diverging opinions on that question.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (mjameswh) [reply temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:133]: Very good point. Dismiss my argument.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:59]: IMO no need to make this const public. I can understand why people need it, but in the world of optional string interning, people can just use a string themselves. If you do want to leave this as public API, please rename to `JAVA_SDK_FAILURE_SOURCE`.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:79]: IMO you don't need all these single-reference static fields, but especially this one. Just inline it in the field below.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:55]: Add javadoc on all public types

File: temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java:119]: I saw a null check later. Do you need to add nullability annotations?

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:101]: I think in our other SDKs this is expected to use the data/payload converter from the parameter

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:120]: > so that users may more easily redefine only the "decision chain", while reusing the encoded_attributes and stack trace bits below.

Nah, IMO these failure converters need to be wrapped not inherited IMO. I'd consider making this whole class "final" but that is a bit much :-). IMO users need to delegate to other failure converters and their entire behavior or not.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:102]: I think you should use a static inner pojo that represents this attribute as a Jackson type for clearer conversion.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:116]: This should be configurable. I totally see users who may want to encode normal payloads and leave Failures unencoded. Pretty much have what they have currently.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:244]: Could you please give these private encode/decode methods more specific names? The amount of overloads drives me a bit crazy on this piece and I don't think it's really needed here, because they are private. I think it will be more readable if it's `encodePayload` and `encodePayloads`

File: temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java:119]: I would suggest marking this method non-null, throw on failure=null and add a javadoc about it. 

I even may see some value for exceptionToFailure to support null. While it shouldn't, but we have situations where we explicitly send encoded "null" Payload, for example as a workflow or activity result.

But there is no reason for a decoding method to support null and I think it will make the contract of this method unclear. Callers of this method should be dealing with null and avoid calling it.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:92]: > The newly created instance will use the default failure converter

Nitpicking, but this method returns `DefaultDataConverter`, so this reads strange and feels like a duplication. I think what you try to say is said on the line above: "populated with the default list of payload converters."

Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:92]:
```suggestion
   * <p>The instance is configured to not encode
```

Nitpicking, but this method returns DefaultDataConverter, so this reads strange and feels like a duplication. I think what you try to say is said on the line above: "populated with the default list of payload converters."

File: temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:111]: nit: leave line 110 and 112 next to each other, they work together :)

File: temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:106]: You may prefer `@link` over `@code` especially when you are linking inside the same class and don't try express "calling a method on a VARIABLE". This will make it much nicer for users to travel around the javadocs, as it will be clickable.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:138]: Let's make it just `withFailureConverter`. In PayloadConverter that `Override` word was added to show that the passed converters with not replace the full collection, but only the relevant "overrode"e elements. I don't see any reason to add it here, it will only make the comprehension of this method harder.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:116]: +1 this is how go failure converters work

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:79]: I think giving cryptic strings a name and a separate place that makes it easier to write a javadoc next to it is a good practice for documentation purposes. I would leave it like this.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:101]: concur, I would expect a custom converter here.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:116]: Oh I see it is configurable it is just in the `encodeFailure` function

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:120]: Agree with Chad. It's definitely not a piece a user should ever customize. If users really want to implement their CORE logic of FailureConvert - they should be forced to implement their own from scratch and take full responsibility for it. Let's not provide an easy way to hijack the core-core logic as it's intended to be overridable. It really is not.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:93]: This sends an incorrect message to users. I think we should actually remove this doc completely to don't give wrong ideas.
Instead, on CodecConverter we should provide information that it takes care of encoding failures if configured to do so with the right constructor flag.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:138]: > I think we should also add a VERY load and clear doc that most users don't need this method. And something like "If you think about encoding the Failures, see CodecPayloadConverter". Users SHOULDN'T typically use this method for failure encoding.

Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DefaultDataConverter.java:138]: Agree. I can't think of a good reason anyone needs a failure converter and since its inception a few months ago I haven't heard of one use of it in the wild (which is why we've been fine for so long not exposing failure conversion semantics to users until now).

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (mjameswh) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:101]: The [spec says](https://github.com/temporalio/api/blob/65015c07229047e82645923cc16285548e642794/temporal/api/failure/v1/message.proto#L97) it should be using the default data converter.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (mjameswh) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:120]: Great! Then I'm actually making this class final. There is just no reason to extend this class rather than implements the interface.

Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:120]: I think that's a bit much unless we change `DefaultDataConverter` to be final too. I think we can discourage inheritance without preventing it. While I generally can understand the final-by-default stance, that is almost a project-wide design decision and I think we generally aren't adding final to all of our classes (@Spikhalskiy correct me if I'm wrong)

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java]: I see in Go and Python that we use the converter given to the default failure converter to convert this two-field JSON object. Can you check TypeScript? If there is a difference we can discuss how to reconcile. For the most part it makes little sense to use a non-default converter I know, but we already have it given to us so why not.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java
Comment (mjameswh) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java]: After verification, `exceptionToFailure` is never called with a null exception. I added `@Nonnull` on that one too.

**Comment by mjameswh** [comment]
I just found out that state machines create a lot of `Failure` directly, without going through `FailureConverter`. I will probably have to fix those too. Converting back to draft in the mean time.

**Comment by mjameswh** [comment]
> I just found out that state machines create a lot of `Failure` directly, without going through `FailureConverter`. I will probably have to fix those too. Converting back to draft in the mean time.

After discussion with @Spikhalskiy on various scenarios, we agree that it is preferable _not_ to modify the way state machines make up `Failure` themselves.

That means that those Failures will not be encrypted, but that seems acceptable since those Failures themselves do not contain any sensible information by themselves (no stack trace, messages are constant from the SDK). In case where the failure's cause might contain sensible data, that cause will come directly from the corresponding history event, and thus, will still be encoded anyway.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (mjameswh) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java]: TypeScript also use the provided payload converter. Modified the Java code to match.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:242]: I think you should use a static inner pojo that represents this attribute as a Jackson type for clearer conversion and field naming.

(restating previous comment since it was outdated)

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:72]: This is not that case of cryptic string used elsewhere. This is just a string that's a single parameter to a regex creation below. If Java had auto-regex strings this wouldn't be a separate field. Not sure it makes sense to have every static final regex pattern be two fields.

```java
  /** Used to parse a stack trace line. */
  private static final String TRACE_ELEMENT_REGEXP =
      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)";

  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(TRACE_ELEMENT_REGEXP);
```

Provides 0 value over:

```java
  /** Used to parse a stack trace line. */
  private static final Pattern TRACE_ELEMENT_PATTERN = Pattern.compile(
      "((?<className>.*)\\.(?<methodName>.*))\\(((?<fileName>.*?)(:(?<lineNumber>\\d+))?)\\)");
```

**Review by cretz — COMMENTED** [review]
2 minor comments

**Comment by Spikhalskiy** [comment]
> > I just found out that state machines create a lot of `Failure` directly, without going through `FailureConverter`. I will probably have to fix those too. Converting back to draft in the mean time.
> 
> After discussion with @Spikhalskiy on various scenarios, we agree that it is preferable _not_ to modify the way state machines make up `Failure` themselves.
> 
> That means that those Failures will not be encrypted, but that seems acceptable since those Failures themselves do not contain any sensible information by themselves (no stack trace, messages are constant from the SDK). In case where the failure's cause might contain sensible data, that cause will come directly from the corresponding history event, and thus, will still be encoded anyway.

Yeah. Also to add to this decision and reasoning, Failures created by State Machines can be approached like History Events. 
For example, a failure created by an Activity State machine from `ActivityTaskCancelled` is unencrypted, but the correspondent Event is also unencrypted already. And the only information in the failure comes from this History Event.

The only exception is the Local Activity State Machine which creates its Failure "out of thin air" and publishes it into a marker. But the type and extent of the data published into this Failure is exactly the same as present in Regular Activity's lifecycle history event and mimicks it. So it's only reasonable to don't encode this Failure, as raw history event fields are unencrypted. The underlying actual cause will be encrypted though.

File: temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java:116]: Let's rename to `encodeFailures`. `default attributes` is a waaaay to internal-api level of naming to be clear for users.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java:129]: I believe we don't capitalize the first letters of param sentences. Also, it's any throwable, not just an exception.

File: temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java:122]: The return value can and should be changed to a more specific TemporalFailure, will give much greater peace of mind for the calling code.

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:78]: RuntimeFailure -> TemporalFailure

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:93]: RuntimeFailure -> TemporalFailure

File: temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/failure/DefaultFailureConverter.java:83]: This block will become unconditional once you make `RuntimeFailure -> TemporalFailure` change.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java:842]: nit: I would just use field directly, it belongs to this object :)

File: temporal-sdk/src/test/java/io/temporal/common/converter/FailureConverterTest.java
Comment (Spikhalskiy) [inline temporal-sdk/src/test/java/io/temporal/common/converter/FailureConverterTest.java:65]: This test belongs under io.temporal.workflow
It's an integration test, all integration tests are there.
We put tests into the package of the tested class if it's a unit test.
Also, a short description will be very appreciated by the future editor. What situations this test covers and fixates?

File: temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java
Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java:122]: I think _technically_ we're allowing users to convert these to/from any exception they want

Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java]: It doesn't make much sense. Any exception other than TemporalFailure in workflow code will cause a workflow task failure, while this thing is persisted in the history already. I don't see any legitimate usecase for non-TemporalFailure to be here.
A user that will try to deserialize a Failure into non-TemporalFailure will face issues down the line.

Comment (mjameswh) [reply temporal-sdk/src/main/java/io/temporal/common/converter/DataConverter.java]: I agree. Other exceptions should be wrapped into some `TemporalFailure`, so that makes sense as the return value from both `DataConverter.failureToException` and `FailureConverter.failureToException`.

**Review by cretz — APPROVED** [review]
LGTM, but please wait for @Spikhalskiy approval before merging.

**Review by Spikhalskiy — APPROVED** [review]

**Comment by Spikhalskiy** [comment]
Great job, that came together rather neatly.

**Pull request merged**
